vulcanai2/models/basenetwork.py

```diff
@@ -235,11 +232,26 @@ def device(self):
     @device.setter
     def device(self, device):
         """
+        Network device setter.
+
         If the user specifies invalid device id, raises
         RuntimeError for invalid device ordinal.
+
+        Parameters
+        ----------
+        device : str or torch.device
+            The device to transfer network to.
+
         """
-        self.network.to(device=device)
-    
+        device = torch.device(device if torch.cuda.is_available() else 'cpu')
+        if self.network:
+            self.network.to(device=device)
+        if self.optim:
+            for state in self.optim.state.values():
```

Comment:
If we have an input network that is on a separate device, and we set the optima parameters to the current device, won't that return an error?

Reply:
Ok I can confirm that this crashes if one of the inputs is on another device. It only works if they are all on the same device

Reply:
Also I think we should handle nn.DataParallel for the users if possible